Validate actual file properties such as `checksum`

tom-tan opened the pull request:
This request fixes #145 by checking that:
- the actual files and directories exist
- the checksum and size of actual file are same as expected values

kinow reviewed (commented):
I think it needs a `make format` or `make cleanup`? Linter failures in the actions.

tom-tan commented:
Done!

mr-c commented on cwltest/utils.py:
```diff
@@ -73,6 +76,21 @@ def compare_location(expected, actual):
         return
     if actual.get("class") == "Directory":
         actual[comp] = actual[comp].rstrip("/")
+        exist_fun = os.path.isdir  # type: Callable[[str], bool]
```
Can you add a test case involving a Directory?

tom-tan replied:
Added!

mr-c commented on cwltest/utils.py:
```diff
@@ -73,6 +76,21 @@ def compare_location(expected, actual):
+    if "path" in actual:
+        path = urllib.parse.urlparse(actual["path"]).path
+    else:
+        path = urllib.parse.urlparse(actual["location"]).path
```
And another test involving `location`?

tom-tan replied:
I added a test for `Directory` with `location`.

mr-c reviewed (changes requested):
Thank you for this!

Some of the new lines need additional tests

Also, it seems we have a Python 3.6 issue now, but that is not the fault of the PR
https://github.com/common-workflow-language/cwltest/actions/runs/3066834974/jobs/4952455385#step:7:291

mr-c reviewed (changes requested):
Thanks for the additional test; There are 5 more lines lacking test coverage as shown on https://github.com/common-workflow-language/cwltest/pull/146/files or via `make diff-cover`

> cwltest/utils.py (86.8%): Missing lines 89,140,150,162,166

tom-tan commented:
I fixed `compare_checksum` and `compare_size` to check the values in the output object and the actual value on disk in addition to checking the expected values and the actual values.
I also added unittests to cover these cases.

tom-tan commented:
ping!

adamnovak commented:
@mr-c If cwltest merges this, it would make it a lot easier to fix https://github.com/DataBiosphere/toil/issues/4210 in Toil. We think Toil is mishandling empty directories in a way cwltest can't currently catch.

mr-c commented:
@tom-tan thank you for starting this; sorry for dropping it!
@adamnovak thanks for the reminder!

mr-c commented:
@adamnovak https://github.com/common-workflow-language/cwltest/releases/tag/2.3.20230527113600 has been released with this functionality
